core/utils/memory.hpp

```diff
@@ -255,86 +255,66 @@ class allocator_array_deleter {
   size_t size_;
 };
 
-struct noop_deleter {
-  template<typename T>
-  void operator()(T* /*unused*/) {}
-};
-
-template<typename T>
-struct managed_deleter : util::noncopyable {
- public:
-  using value_type = T;
-  using pointer = T*;
-
-  explicit managed_deleter(pointer ptr = nullptr) noexcept : ptr_(ptr) {}
+struct Managed {
+ protected:
+  virtual ~Managed() = default;
 
-  template<typename U,
-           typename = std::enable_if_t<std::is_convertible_v<U*, pointer>, U*>>
-  // cppcheck-suppress noExplicitConstructor
-  managed_deleter(managed_deleter<U>&& rhs) noexcept : ptr_(rhs.ptr_) {
-    rhs.ptr_ = nullptr;
-  }
+ private:
+  friend struct ManagedDeleter;
+
+  // Const because we can allocate and then delete const object
+  //
+  // Also it's possible to make here empty implementation
+  // and it makes usage more convenient but also more error prone
+  // and don't finalize class
+  virtual void Destroy() const = 0;
+};
 
-  managed_deleter(managed_deleter&& rhs) noexcept : ptr_(rhs.ptr_) {
-    rhs.ptr_ = nullptr;
-  }
+template<typename Base>
+struct OnStack final : Base {
+  static_assert(std::is_base_of_v<Managed, Base>);
 
-  managed_deleter& operator=(managed_deleter&& rhs) noexcept {
-    if (this != &rhs) {
-      ptr_ = rhs.ptr_;
-      rhs.ptr_ = nullptr;
-    }
-    return *this;
-  }
+  template<typename... Args>
+  OnStack(Args&&... args) : Base{std::forward<Args>(args)...} {}
 
-  template<typename U,
-           typename = std::enable_if_t<std::is_convertible_v<U*, pointer>, U*>>
-  managed_deleter& operator=(managed_deleter<U>&& rhs) noexcept {
-    ptr_ = rhs.ptr_;
-    rhs.ptr_ = nullptr;
-    return *this;
-  }
+ private:
+  void Destroy() const final {}
+};
 
-  void operator()(const pointer p) noexcept {
-    IRS_ASSERT(!ptr_ || p == ptr_);
-    delete ptr_;
-  }
+template<typename Base>
+struct OnHeap final : Base {
+  static_assert(std::is_base_of_v<Managed, Base>);
 
-  pointer get() const noexcept { return ptr_; }
+  template<typename... Args>
+  OnHeap(Args&&... args) : Base{std::forward<Args>(args)...} {}
 
  private:
-  template<typename U>
-  friend struct managed_deleter;
+  void Destroy() const final { delete this; }
+};
 
-  pointer ptr_;
-};  // managed_deleter
+struct ManagedDeleter {
+  void operator()(const Managed* p) noexcept {
+    IRS_ASSERT(p != nullptr);  // std::unique_ptr doesn't call dtor on nullptr
+    p->Destroy();
```

Comment:
Isn't the call to `Destroy()` here a virtual function call? Should `ManagedDeleter` be templated using the type of the specialized class it manages?

Reply:
It cannot be templated, type of deleter will be different

Reply:
Also it always was virtual, because before my changes was call of virtual dtor